### ISSUE TITLE
Iterator helpers TC39 proposal

### ISSUE THREAD

**Issue by danielbayley** [DanielXMoore/Civet#1301]
What do we think about adding polyfills for [iterator helpers](https://github.com/tc39/proposal-iterator-helpers#readme)—currently at stage 3 of the TC39 process—and their [`async` equivalent](https://github.com/tc39/proposal-async-iterator-helpers#readme), currently at stage 2.

**Comment by bbrk24** [comment]
I wouldn't be against it, but aren't polyfills generally outside the scope of Civet?

**Comment by edemaine** [comment]
I agree that it seems out of the scope, given that this proposal seems to be a standard-library feature instead of a language feature. (Let me know if I misread.) So it seems you that you can just include an existing JS polyfill, and it'll work great with Civet.

**Comment by danielbayley** [comment]
> given that this proposal seems to be a standard-library feature instead of a language feature. So it seems you that you can just include an existing JS polyfill

Ah of course, makes sense. Do you happen to know of a good one?

I did wonder if Civet could possibly have some kind of `[...asyncThing]` sugar for `Array.fromAsync`…

**Comment by edemaine** [comment]
The best polyfill I've found is https://github.com/MadProbe/iterator-helpers-polyfill
It's a bit out-of-date in the async side (see issue 15), but hopefully good enough for your purposes...

I'm not sure we want to redefine `[...asyncThing]` (though I see `Array.fromAsync` is backward-compatible in the sense that it supports regular iterators), but perhaps new notation like `[...*asyncThing]` could trigger this?  Ideally we could also support `[a, ...*b, ...c, ...*d]` via something like `[a, ...Array.fromAsync(b), ...c, ...Array.fromAsync(d)]` (unless we can think up a more efficient compilation).

**Comment by bbrk24** [comment]
`Array::concat` can handle multiple arguments, and is way faster than multiple spreads: [JSBenchmark](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJsMU5kMlRoVm41NHRkclhrR3ZjY20iLCJjb2RlIjoiREFUQS5hLmNvbmNhdChEQVRBLmIsIERBVEEuYykiLCJuYW1lIjoiQ2FsbCAuY29uY2F0IiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6IjhMdU9FSlJUanFXMUo0TTVLMVlTLSIsImNvZGUiOiJbLi4uREFUQS5hLCAuLi5EQVRBLmIsIC4uLkRBVEEuY10iLCJuYW1lIjoiVXNlIHNwcmVhZCBvcGVyYXRvciIsImRlcGVuZGVuY2llcyI6W119XSwiY29uZmlnIjp7Im5hbWUiOiJCYXNpYyBleGFtcGxlIiwicGFyYWxsZWwiOnRydWUsImdsb2JhbFRlc3RDb25maWciOnsiZGVwZW5kZW5jaWVzIjpbXX0sImRhdGFDb2RlIjoicmV0dXJuIHsgYTogWy4uLkFycmF5KDEwMDApLmtleXMoKV0sIGI6IFsuLi5BcnJheSgxMDAwKS5rZXlzKCldLCBjOiBbLi4uQXJyYXkoMTAwMCkua2V5cygpXSB9In19) .concat is about 5x faster for me (Firefox on Windows). So, the compilation would be `[a].concat(Array.fromAsync(b), c, Array.fromAsync(d))`.
